ref(actions): allow more flexibility on cached states usage

gustavovalverde opened the pull request:
## Motivation

We've added the possibility to deploy instances with a cached state attached, but we might also want to deploy them without it, or change the cached state disk preference depending on the github source, while not affecting our actual deployment process.

## Solution

- Allow to select a cached state based on:
  - Type of cached disk to use (`tip` or `checkpoint`)
  - Source (`main` branch vs `PR` or `any` other available)
  - Network
- Allow to deploy a node without cached state
- Search for branch/PR reference instead of commit (as this is an use case that we haven't used for a while)
- Improve the disk selection logic, by allowing a preference for cached states from the `main` branch, without disallowing finding other available disks

### Tests

- Deployed to Testnet with a cached state: https://github.com/ZcashFoundation/zebra/actions/runs/11162158270
- Deployed to Testnet without a cached state: https://github.com/ZcashFoundation/zebra/actions/runs/11163638750 (get disk is skipped and instance is deployed without a cached state attached).

### Follow-up Work

Support the `matrix` use case when releasing a new Zebra version: https://github.com/ZcashFoundation/zebra/blob/2552c3c94bee3d24db7ae75ae67663400c6fcad6/.github/workflows/cd-deploy-nodes-gcp.yml#L233 

### PR Author's Checklist

<!-- If you are the author of the PR, check the boxes below before making the PR
ready for review. -->

- [x] The PR name will make sense to users.
- [x] The PR provides a CHANGELOG summary.
- [x] The solution is tested.
- [x] The documentation is up to date.
- [x] The PR has a priority label.

### PR Reviewer's Checklist

<!-- If you are a reviewer of the PR, check the boxes below before approving it. -->

- [ ] The PR Author's checklist is complete.
- [ ] The PR resolves the issue.

gustavovalverde committed:
ref(actions): allow more flexibility on cached states usage

gustavovalverde committed:
chore: improve message

gustavovalverde committed:
fix(actions): deploy single instances even if no cached state is needed

upbqdn commented on .github/workflows/cd-deploy-nodes-gcp.yml:
```diff
@@ -250,14 +273,14 @@ jobs:
+          DISK_PARAMS="name=${DISK_NAME},device-name=${DISK_NAME},size=400GB,type=pd-ssd"
           if [ -n "${{ env.CACHED_DISK_NAME }}" ]; then
             DISK_PARAMS+=",image=${{ env.CACHED_DISK_NAME }}"
+          elif [ ${{ !inputs.no_cached_disk && github.event_name == 'workflow_dispatch' }} ]; then
```
Shouldn't this be without the exclamation mark?

```suggestion
          elif [ ${{ inputs.no_cached_disk && github.event_name == 'workflow_dispatch' }} ]; then
```

upbqdn commented on .github/workflows/cd-deploy-nodes-gcp.yml:
```diff
@@ -326,7 +349,8 @@ jobs:
       id-token: 'write'
-    if: github.event_name == 'workflow_dispatch'
+    # Run even if we don't need a cached disk, but only when triggered by a workflow_dispatch
+    if: ${{ !failure(needs.get-disk-name) && github.event_name == 'workflow_dispatch' }}
```
It looks like we can't use `failure` this way: https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/evaluate-expressions-in-workflows-and-actions#failure.

gustavovalverde replied:
Seems like I was sleepy when doing this

upbqdn commented on .github/workflows/cd-deploy-nodes-gcp.yml:
```diff
@@ -363,10 +387,12 @@ jobs:
+          DISK_PARAMS="name=${DISK_NAME},device-name=${DISK_NAME},size=400GB,type=pd-ssd"
           if [ -n "${{ env.CACHED_DISK_NAME }}" ]; then
             DISK_PARAMS+=",image=${{ env.CACHED_DISK_NAME }}"
+          elif [ ${{ !inputs.no_cached_disk && github.event_name == 'workflow_dispatch' }} ]; then
```
Same as above

```suggestion
          elif [ ${{ inputs.no_cached_disk && github.event_name == 'workflow_dispatch' }} ]; then
```

upbqdn commented on .github/workflows/scripts/gcp-get-cached-disks.sh:
```diff
@@ -20,7 +20,7 @@ echo "Extracting local state version..."
 echo "STATE_VERSION: ${LOCAL_STATE_VERSION}"
 
-# Function to find a cached disk image based on the git pattern (commit, main, or any branch)
+# Function to find a cached disk image based on the git pattern (branch, main, or any branch)
```
###### Optional

I think we can keep the "commit" here.

```suggestion
# Finds a cached disk image based on the given git pattern (commit, main, or any branch)
```

gustavovalverde replied:
This is no longer looking for a specific commit but a github reference, which in most cases is the PR branch, or the main branch. https://github.com/ZcashFoundation/zebra/pull/8908/files/2552c3c94bee3d24db7ae75ae67663400c6fcad6#diff-3fe6320dbec0ab834aa0e833b666416668b3538ab84b4ce0d90b817867d6d208L50

upbqdn commented on .github/workflows/scripts/gcp-get-cached-disks.sh:
```diff
@@ -43,31 +43,29 @@ find_cached_disk_image() {
+        CACHED_DISK_NAME=$(find_cached_disk_image "main-[0-9a-f]+" "main branch")
+    # Else, try to find a cached disk image from the current branch (or PR)
+    else
+        CACHED_DISK_NAME=$(find_cached_disk_image ".+-${GITHUB_REF}" "branch")
```
Shouldn't we test if `GITHUB_REF` isn't empty before we pass it to `find_cached_disk_image`?

gustavovalverde replied:
I think the only benefit will be that we'd be able to `echo` an error in the logs. But if the `GITHUB_REF` is missing, then there's a bunch of other things that will also fail in other workflows, as there's not supposed to be a commit without a reference.

Unless you're thinking in other future scenarios with different ways to call this script.

upbqdn commented on .github/workflows/scripts/gcp-get-cached-disks.sh:
```diff
@@ -3,9 +3,9 @@
 #
 # If there are multiple disks:
-# - prefer images generated from the same commit, then
+# - prefer images generated from the same branch, then
```
```suggestion
# - if `PREFER_MAIN_CACHED_STATE` is "true", then select an image from the `main` branch, else
```

upbqdn commented on .github/workflows/scripts/gcp-get-cached-disks.sh:
```diff
@@ -3,9 +3,9 @@
 # If there are multiple disks:
-# - prefer images generated from the same commit, then
+# - prefer images generated from the same branch, then
 # - if prefer_main_cached_state is true, prefer images from the `main` branch, then
```
```suggestion
# - try to find a cached disk image from the current branch (or PR), else
```

upbqdn commented on .github/workflows/scripts/gcp-get-cached-disks.sh:
```diff
@@ -3,9 +3,9 @@
+# - prefer images generated from the same branch, then
 # - if prefer_main_cached_state is true, prefer images from the `main` branch, then
-# - use any images from any other branch or commit.
+# - use any images from any other branch or branch.
```
```suggestion
# - try to find an image from any branch.
```

gustavovalverde committed:
rev: apply suggestions from code review

Co-authored-by: Marek <mail@marek.onl>

gustavovalverde committed:
fix: wrong use of `failure()`

upbqdn commented:
@gustavovalverde can you update this branch? I can't do it.

gustavovalverde committed:
chore: remove extra file

upbqdn commented on .github/workflows/sub-find-cached-disks.yml:
```diff
@@ -74,20 +74,28 @@ jobs:
-      # uses sentence case, so we need to downcase ${{ inputs.network }}
+      # uses sentence case, so we need to downcase ${{ inputs.network }}.
+      #
+      # Disk image names in GCP are limited to 63 characters, so we need to limit
```
Can you add a link to further details or describe the other 51 chars?

gustavovalverde replied:
Solved with https://github.com/ZcashFoundation/zebra/pull/8908/commits/ba68afde7ccbe46450d6ea092e5546e0889e4a36

gustavovalverde commented:
@upbqdn this one should be ready to merge

gustavovalverde committed:
chore: `echo` the pattern for easier debugging

upbqdn commented on .github/workflows/scripts/gcp-get-cached-disks.sh:
```diff
@@ -36,38 +36,36 @@ find_cached_disk_image() {
-        # Else, try to find one from any branch
-        else
+    # Check if main branch images are preferred
+    if [[ "${PREFER_MAIN_CACHED_STATE}" == "true" ]]; then
```
If `PREFER_MAIN_CACHED_STATE` is `true` and `find_cached_disk_image` doesn't find anything, then the script fails. Is this the intended behavior? I can open a quick PR to handle this case since I've already given it a thought.

upbqdn replied:
- This is what I had in mind: #8936.

gustavovalverde replied:
Solved with https://github.com/ZcashFoundation/zebra/pull/8908/commits/50cd029a75d4e8dad88e76b3fc233a173b35e0cd

gustavovalverde committed:
chore: add extra details to image naming convention

Addresses https://github.com/ZcashFoundation/zebra/pull/8908#discussion_r1797679586

gustavovalverde committed:
ref(actions): use a better logic for disk image selection

This supersedes https://github.com/ZcashFoundation/zebra/pull/8936/ using a different approach with `${VAR:-value}`
